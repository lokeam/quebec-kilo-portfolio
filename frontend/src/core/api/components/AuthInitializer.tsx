import { useAuth0 } from '@auth0/auth0-react';
import { useEffect } from 'react';
import * as Sentry from '@sentry/react';
import { saveAuth0Token } from '@/core/api/utils/auth.utils';

/**
 * AuthInitializer is a crucial component that bridges Auth0 authentication
 * with the API layer. Ensures that every API request has access to
 * the latest auth token.
 *
 * Why this exists:
 * 1. Our API layer needs auth tokens for requests
 * 2. Auth0 manages these tokens
 * 3. We need a way to get fresh tokens when making requests
 * 4. This component provides that connection
 *
 * Component Hierarchy:
 * ```
 * <Auth0Provider>
 *   <AuthInitializer>     <--- We are here
 *     <QueryClientProvider>
 *       <App />
 *     </QueryClientProvider>
 *   </AuthInitializer>
 * </Auth0Provider>
 * ```
 *
 * @example
 * ```typescript
 * // main.tsx
 * root.render(
 *   <Auth0Provider {...auth0Config}>
 *     <AuthInitializer>
 *       <App />
 *     </AuthInitializer>
 *   </Auth0Provider>
 * );
 *
 * // Later used in API calls:
 * const response = await axiosInstance.get('/api/data'); // Token automatically included
 * ```
 *
 * @param props.children - Child components that need access to authenticated API calls
 */
export function AuthInitializer({ children }: { children: React.ReactNode }) {
  const { getAccessTokenSilently, user, isAuthenticated } = useAuth0();

  useEffect(() => {
    // Save Auth0's getAccessTokenSilently fn so that we may use tokens anywhere.
    // Needed bc getAccessTokenSilently only works in React components.
    // Fn wrapper used to pass different options later if needed.
    saveAuth0Token(() => getAccessTokenSilently());
  }, [getAccessTokenSilently]);

  // Set Sentry user context when authentication state changes
  useEffect(() => {
    if (isAuthenticated && user) {
      // Set user context in Sentry for error tracking
      Sentry.setUser({
        id: user.sub,
        email: user.email,
        username: user.name || user.email,
      });

      // Add user properties for better errors
      Sentry.setContext('user', {
        email: user.email,
        name: user.name,
        email_verified: user.email_verified,
        sub: user.sub,
      });
    } else {
      // Clear user context when not authenticated
      Sentry.setUser(null);
    }
  }, [isAuthenticated, user]);

  return <>{children}</>;
}

/**
 * Auth Initializer Component
 *
 * For API standards and best practices, see:
 * @see {@link ../../docs/api-standards.md}
 */
