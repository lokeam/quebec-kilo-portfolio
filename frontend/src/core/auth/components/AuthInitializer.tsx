import { useAuth0 } from '@auth0/auth0-react';
import { useEffect } from 'react';
import * as Sentry from '@sentry/react';
import { saveAuth0Token } from '@/core/api/utils/auth.utils';

/**
 * AuthInitializer is a crucial component that bridges Auth0 authentication
 * with the API layer. Ensures that every API request has access to
 * the latest auth token.
 *
 * Why this exists:
 * 1. Our API layer needs auth tokens for requests
 * 2. Auth0 manages these tokens
 * 3. We need a way to get fresh tokens when making requests
 * 4. This component provides that connection
 *
 * Component Hierarchy:
 * ```
 * <Auth0Provider>
 *   <AuthInitializer>     <--- We are here
 *     <QueryClientProvider>
 *       <App />
 *     </QueryClientProvider>
 *   </AuthInitializer>
 * </Auth0Provider>
 * ```
 *
 * @example
 * ```typescript
 * // main.tsx
 * root.render(
 *   <Auth0Provider {...auth0Config}>
 *     <AuthInitializer>
 *       <App />
 *     </AuthInitializer>
 *   </Auth0Provider>
 * );
 *
 * // Later used in API calls:
 * const response = await axiosInstance.get('/api/data'); // Token automatically included
 * ```
 *
 * @param props.children - Child components that need access to authenticated API calls
 */
export function AuthInitializer({ children }: { children: React.ReactNode }) {
  const { getAccessTokenSilently, user, isAuthenticated, isLoading } = useAuth0();

  useEffect(() => {
    // Save Auth0's getAccessTokenSilently fn so that we may use tokens anywhere.
    // Needed bc getAccessTokenSilently only works in React components.
    // Fn wrapper used to pass different options later if needed.
    saveAuth0Token(() => getAccessTokenSilently());
  }, [getAccessTokenSilently]);

  // Ensure theme consistency during Auth0 callback processing
  useEffect(() => {
    // Check if we're in an Auth0 callback state
    const isAuth0Callback = window.location.search.includes('code=') && window.location.search.includes('state=');

    if (isAuth0Callback) {
      // Apply theme immediately during Auth0 callback to prevent white flash
      try {
        const storedTheme = localStorage.getItem('qko-theme-storage');
        let themeMode = 'light';

        if (storedTheme) {
          try {
            const parsed = JSON.parse(storedTheme);
            themeMode = parsed.mode || 'light';
          } catch {
            console.warn('Failed to parse stored theme during Auth0 callback');
          }
        }

        const html = document.documentElement;
        const body = document.body;
        const root = document.getElementById('root');

        if (themeMode === 'dark') {
          html.classList.add('dark');
          html.style.backgroundColor = 'hsl(222.2 47.4% 11.2%)';
          html.style.color = 'hsl(210 40% 98%)';
          if (body) {
            body.style.backgroundColor = 'hsl(222.2 47.4% 11.2%)';
            body.style.color = 'hsl(210 40% 98%)';
          }
          if (root) {
            root.style.backgroundColor = 'hsl(222.2 47.4% 11.2%)';
            root.style.color = 'hsl(210 40% 98%)';
          }
        } else {
          html.classList.remove('dark');
          html.style.backgroundColor = 'hsl(0 0% 100%)';
          html.style.color = 'hsl(222.2 47.4% 11.2%)';
          if (body) {
            body.style.backgroundColor = 'hsl(0 0% 100%)';
            body.style.color = 'hsl(222.2 47.4% 11.2%)';
          }
          if (root) {
            root.style.backgroundColor = 'hsl(0 0% 100%)';
            root.style.color = 'hsl(222.2 47.4% 11.2%)';
          }
        }
      } catch (error) {
        console.warn('Failed to apply theme during Auth0 callback:', error);
      }
    }
  }, [isLoading]);

  // Set Sentry user context when authentication state changes
  useEffect(() => {
    if (isAuthenticated && user) {
      // Set user context in Sentry for error tracking
      Sentry.setUser({
        id: user.sub,
        email: user.email,
        username: user.name || user.email,
      });

      // Add user properties for better errors
      Sentry.setContext('user', {
        email: user.email,
        name: user.name,
        email_verified: user.email_verified,
        sub: user.sub,
      });
    } else {
      // Clear user context when not authenticated
      Sentry.setUser(null);
    }
  }, [isAuthenticated, user]);

  return <>{children}</>;
}