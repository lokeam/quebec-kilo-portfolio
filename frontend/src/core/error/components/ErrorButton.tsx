import { Button } from '@/shared/components/ui/button/button';
import { cn } from '@/shared/components/ui/utils';

interface ErrorButtonProps {
  onClick: () => void;
  label?: string;
  disabled?: boolean;
  'aria-label'?: string;
  className?: string;
  onKeyDown?: (event: React.KeyboardEvent) => void;
};

export const ErrorButton = ({
  onClick,
  label = 'Try Again',
  disabled = false,
  'aria-label': ariaLabel,
  className = '',
  onKeyDown,
} : ErrorButtonProps) => {
  console.log('ErrorButton');
  return (
    <Button
      variant="default"
      onClick={onClick}
      onKeyDown={onKeyDown}
      disabled={disabled}
      aria-label={ariaLabel}
      className={cn(
        "min-w-[100px] m-4",
        className
      )}
    >
      {label}
    </Button>
  );
};


