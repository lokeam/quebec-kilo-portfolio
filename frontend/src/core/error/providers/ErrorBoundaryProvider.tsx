// Sentry error boundary
import * as Sentry from '@sentry/react';
import { useLocation } from 'react-router-dom';
import { categorizeError } from '@/core/error/categorization';

import { DefaultErrorFallbackPage } from '@/core/error/pages/DefaultErrorFallbackPage';

// Note: Removed Tanstack Query error boundary in favor of Sentry
// import { QueryErrorResetBoundary } from '@tanstack/react-query';


interface ErrorBoundaryProviderProps {
  children: React.ReactNode;
  FallbackComponent?: React.ComponentType<{
    error: Error;
    resetErrorBoundary: () => void;
  }>;
}

/**
 * Global error boundary provider that catches runtime errors and displays fallback UI
 * Based on React Error Boundary pattern: https://react.dev/reference/react/Component#catching-rendering-errors-with-an-error-boundary
 */
export const ErrorBoundaryProvider = ({
  children,
  FallbackComponent = DefaultErrorFallbackPage
}: ErrorBoundaryProviderProps) => {
  const location = useLocation();

  return (
    <Sentry.ErrorBoundary
      fallback={({ error, resetError }) => (
        <FallbackComponent
          error={error as Error}
          resetErrorBoundary={resetError}
        />
      )}
      beforeCapture={(scope, hint) => {
        // Categorize the error
        const error = (hint as { originalException?: Error })?.originalException;
        if (error) {
          const errorContext = categorizeError(error, {
            currentPage: location.pathname,
            timestamp: new Date().toISOString(),
          });

          // Add error categorization tags
          Object.entries(errorContext.tags).forEach(([key, value]) => {
            scope.setTag(key, value);
          });

          // Add error context
          scope.setContext('error_categorization', errorContext.context);
        }

        // Set error level based on priority
        scope.setLevel("error");

        // Add navigation context
        scope.setContext('navigation', {
          currentRoute: location.pathname,
          search: location.search,
          hash: location.hash,
        });

        // Add performance context if available
        if (window.performance) {
          const navigation = performance.getEntriesByType('navigation')[0] as PerformanceNavigationTiming;
          if (navigation) {
            scope.setContext('performance', {
              pageLoadTime: navigation.loadEventEnd - navigation.loadEventStart,
              domContentLoaded: navigation.domContentLoadedEventEnd - navigation.domContentLoadedEventStart,
              firstPaint: performance.getEntriesByName('first-paint')[0]?.startTime,
              firstContentfulPaint: performance.getEntriesByName('first-contentful-paint')[0]?.startTime,
            });
          }
        }
      }}
    >
      {children}
    </Sentry.ErrorBoundary>
  );
};
