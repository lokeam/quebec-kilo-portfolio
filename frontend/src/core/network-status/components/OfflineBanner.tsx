
import { useNetworkStatus } from '@/core/network-status/hooks/useNetworkStatus';
import { WifiOff } from 'lucide-react';


export const OfflineBanner = () => {
  const { isOnline } = useNetworkStatus();

  if (isOnline) return null;

  return (
      <div
        className="offline_banner bg-charcoal relative box-border flex flex-col min-h-min items-center content-center h-full py-4 px-2"
        role="alert"
        style={{
          position: 'fixed',
          top: '0px',
          left: '0px',
          right: '0px',
          width: '100%',
          zIndex: 1000
        }}
      >
        <div className="flex flex-col w-full items-center content-center">
          <div className="w-64 flex flex-row">
            <div className="relative inline-flex items-center justify-center overflow-hidden bg-transparent h-9 w-9 rounded-full">
              <WifiOff color="#fff" size={30} />
          </div>
            <div className="flex flex-col items-center justify-center text-lg font-bold text-center align-middle ml-3">You are offline. Please check your internet connection.</div>
          </div>
        </div>
      </div>
  );
};
