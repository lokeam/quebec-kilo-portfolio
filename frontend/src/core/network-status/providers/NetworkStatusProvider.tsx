import {
  createContext,
  useState,
  useEffect,
  useContext,
  type ReactNode,
} from 'react';

import { NETWORK_STATUS_ERRORS } from '../constants/errorConstants';
import { type NetworkStatusContextType } from '../types/networkStatusType';

// Context for managing network connectivity state across app. Undefined until Provider is mounted.
export const NetworkStatusContext = createContext<NetworkStatusContextType | undefined>(undefined);

/*
  Monitors and broadcasts network connectivity status.
  Must be placed high in component tree to provide status to all components.
*/
export function NetworkStatusProvider({ children}: { children: ReactNode }) {
  // Init state w/ current network status
  const [isOnline, setIsOnline] = useState<boolean>(navigator.onLine);

  useEffect(() => {
    const handleOnline = () => setIsOnline(true);
    const handleOffline = () => setIsOnline(false);

    window.addEventListener('online', handleOnline);
    window.addEventListener('offline', handleOffline);

    // Initial status check
    setIsOnline(navigator.onLine);

    // Cleanup fn to remove event listeners on unmount
    return  () => {
      window.removeEventListener('online', handleOnline);
      window.removeEventListener('offline', handleOffline);
    }
  }, []);

  return (
    <NetworkStatusContext.Provider value= {{ isOnline}} >
      { children }
    </NetworkStatusContext.Provider>
  );
}

/*
  Custom hook to access network status from any component.
  Must be used within a NetworkStatusProvider
*/
export const useNetworkStatus = () => {
  const context = useContext(NetworkStatusContext);
  if (context === undefined) {
    throw new Error(NETWORK_STATUS_ERRORS.PROVIDER_MISSING);
  }

  return context;
}