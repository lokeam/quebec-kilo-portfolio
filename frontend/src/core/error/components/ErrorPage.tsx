import { ErrorPageOneColTemplate } from '@/core/error/components/ErrorPageOneColTemplate';
import { ErrorPageTwoColTemplate } from '@/core/error/components/ErrorPageTwoColTemplate';

export type ErrorPageVariant = '404' | '500';

export interface ErrorPageProps {
  variant: ErrorPageVariant;
  title: string;
  subtext?: string;
  buttonText?: string;
  role?: 'alert' | 'status';
  ariaLive?: 'polite' | 'assertive';
  onButtonClick?: () => void;
  image?: React.ReactNode;
  children?: React.ReactNode;
}

export const ErrorPage: React.FC<ErrorPageProps> = (props) => {
  const { variant, ...rest } = props;
  if (variant === '404') {
    return <ErrorPageOneColTemplate {...rest} />;
  }
  return <ErrorPageTwoColTemplate {...rest} />;
};