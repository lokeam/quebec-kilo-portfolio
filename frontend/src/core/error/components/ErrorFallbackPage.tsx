import type { FallbackProps } from 'react-error-boundary';
import { ErrorButton } from '@/core/error/components/ErrorButton';
import type { ErrorConfig } from '@/core/error/types/error.types';
import { cn } from '@/shared/components/ui/utils';

interface ErrorFallbackPageProps extends FallbackProps {
  config?: ErrorConfig;
};

export const ErrorFallbackPage = ({
  error,
  resetErrorBoundary, // Provided by react-error-boundary
  config
}: ErrorFallbackPageProps) => {

  const handleClick = () => {
    if (config?.onAction) {
      // If custom action exists, only call that
      config.onAction();
    } else {
      // Otherwise use the default reset behavior
      resetErrorBoundary();
    }
  };

  const handleKeyDown = (event: React.KeyboardEvent) => {
    if (event.key === 'Enter' || event.key === ' ') {
       // Prevent page scroll on space
      event.preventDefault();
      if (config?.onAction) {
        config.onAction();
      } else {
        resetErrorBoundary();
      }
    }
  };
  // Log error for monitoring
  console.error('[ErrorBoundary] Error:', error);

  return (
    <div className="container mx-auto max-w-2xl px-4">
      <div className={cn(
        "mt-32 p-8 rounded-lg shadow-lg bg-background",
        "flex flex-col items-center"
      )}>
        <div className="text-center">
          <h2 className={cn(
            "text-3xl font-bold mb-4",
            "text-destructive"
          )}>
            {config?.severity === 'fatal' ? 'Critical Error' : 'Oops!'}
          </h2>

          <h3 className="text-xl font-semibold mb-4">
            {config?.message || 'Something went wrong'}
          </h3>

          <p className="text-muted-foreground mb-6">
            {error?.message || 'We apologize for the inconvenience. Please try again.'}
          </p>

          <ErrorButton
            onClick={handleClick}
            onKeyDown={handleKeyDown}
            label={config?.actionLabel || 'Try Again'}
          />
        </div>
      </div>
    </div>
  );
};