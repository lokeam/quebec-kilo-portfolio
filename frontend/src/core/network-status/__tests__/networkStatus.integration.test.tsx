import { render, screen, act, waitFor } from '@testing-library/react';
import { NetworkStatusProvider } from '../providers/NetworkStatusProvider';
import { OfflineBanner } from '@/core/network-status/components/OfflineBanner';
import { vi } from 'vitest';

describe('Network Status Integration', () => {
  const renderApp = () => {
    return render(
      <NetworkStatusProvider>
        <div data-testid="app-content" style={{ height: '100vh' }}>
          <OfflineBanner />
          <div>Main Content</div>
        </div>
      </NetworkStatusProvider>
    );
  };

  beforeEach(() => {
    // Reset any previous network status
    window.setNetworkStatus(true);
  });

  it('should handle real browser online/offline events', async () => {
    renderApp();

    // Initially no banner
    expect(screen.queryByRole('alert')).not.toBeInTheDocument();

    // Simulate real browser offline event
    act(() => {
      window.dispatchEvent(new Event('offline'));
    });

    // Wait for the banner to appear
    await waitFor(() => {
      expect(screen.getByRole('alert')).toBeInTheDocument();
    });

    const banner = screen.getByRole('alert');
    expect(banner).toHaveTextContent('You are offline. Please check your internet connection.');

    // Verify banner positioning and layout
    expect(banner).toHaveStyle({
      position: 'fixed',
      top: '0px',
      left: '0px',
      right: '0px',
      width: '100%',
      zIndex: '1000',
    });

    // Ensure no content shift when banner appears
    const appContent = screen.getByTestId('app-content');
    const initialHeight = appContent.clientHeight;

    // Simulate real browser online event
    act(() => {
      window.dispatchEvent(new Event('online'));
    });

    // Wait for banner to disappear
    await waitFor(() => {
      expect(screen.queryByRole('alert')).not.toBeInTheDocument();
    });

    // Verify no layout shift occurred
    expect(appContent.clientHeight).toBe(initialHeight);
  });

  it('should propagate state changes through component tree', async () => {
    const { container } = renderApp();

    // Take initial snapshot
    expect(container).toMatchSnapshot();

    // Go offline
    await act(async () => {
      window.setNetworkStatus(false);
    });

    // Take offline snapshot
    expect(container).toMatchSnapshot();

    // Go back online
    await act(async () => {
      window.setNetworkStatus(true);
    });

    // Take final snapshot
    expect(container).toMatchSnapshot();
  });

  it('should handle rapid network status changes', async () => {
    renderApp();

    await act(async () => {
      // Simulate network flapping
      window.dispatchEvent(new Event('offline'));
      window.dispatchEvent(new Event('online'));
      window.dispatchEvent(new Event('offline'));
    });

    // Should end up showing banner in offline state
    expect(screen.getByRole('alert')).toBeInTheDocument();
  });

  it('should cleanup event listeners', () => {
    const removeEventListenerSpy = vi.spyOn(window, 'removeEventListener');
    const { unmount } = renderApp();

    unmount();

    expect(removeEventListenerSpy).toHaveBeenCalledWith('online', expect.any(Function));
    expect(removeEventListenerSpy).toHaveBeenCalledWith('offline', expect.any(Function));
  });
});