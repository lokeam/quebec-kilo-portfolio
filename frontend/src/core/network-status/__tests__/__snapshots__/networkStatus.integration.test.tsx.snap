// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Network Status Integration > should propagate state changes through component tree 1`] = `
<div>
  <div
    data-testid="app-content"
    style="height: 100vh;"
  >
    <div>
      Main Content
    </div>
  </div>
</div>
`;

exports[`Network Status Integration > should propagate state changes through component tree 2`] = `
<div>
  <div
    data-testid="app-content"
    style="height: 100vh;"
  >
    <div
      class="offline_banner bg-charcoal relative box-border flex flex-col min-h-min items-center content-center h-full py-4 px-2"
      role="alert"
      style="position: fixed; top: 0px; left: 0px; right: 0px; width: 100%; z-index: 1000;"
    >
      <div
        class="flex flex-col w-full items-center content-center"
      >
        <div
          class="w-64 flex flex-row"
        >
          <div
            class="relative inline-flex items-center justify-center overflow-hidden bg-transparent h-9 w-9 rounded-full"
          >
            <svg
              class="lucide lucide-wifi-off"
              fill="none"
              height="30"
              stroke="#fff"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="30"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M12 20h.01"
              />
              <path
                d="M8.5 16.429a5 5 0 0 1 7 0"
              />
              <path
                d="M5 12.859a10 10 0 0 1 5.17-2.69"
              />
              <path
                d="M19 12.859a10 10 0 0 0-2.007-1.523"
              />
              <path
                d="M2 8.82a15 15 0 0 1 4.177-2.643"
              />
              <path
                d="M22 8.82a15 15 0 0 0-11.288-3.764"
              />
              <path
                d="m2 2 20 20"
              />
            </svg>
          </div>
          <div
            class="flex flex-col items-center justify-center text-lg font-bold text-center align-middle ml-3"
          >
            You are offline. Please check your internet connection.
          </div>
        </div>
      </div>
    </div>
    <div>
      Main Content
    </div>
  </div>
</div>
`;

exports[`Network Status Integration > should propagate state changes through component tree 3`] = `
<div>
  <div
    data-testid="app-content"
    style="height: 100vh;"
  >
    <div>
      Main Content
    </div>
  </div>
</div>
`;
